packages/async-flow/src/weak-bijection.js

```diff
@@ -0,0 +1,127 @@
+import { Fail } from '@endo/errors';
+import { M } from '@endo/patterns';
+import { Far } from '@endo/pass-style';
+import { toPassableCap } from '@agoric/vow';
+import { makeEphemera } from './ephemera.js';
+
+const WeakBijectionI = M.interface('WeakBijection', {
+  reset: M.call().returns(),
+  init: M.call(M.any(), M.any()).returns(),
+  hasGuest: M.call(M.any()).returns(M.boolean()),
+  hasHost: M.call(M.any()).returns(M.boolean()),
+  has: M.call(M.any(), M.any()).returns(M.boolean()),
+  guestToHost: M.call(M.any()).returns(M.any()),
+  hostToGuest: M.call(M.any()).returns(M.any()),
+});
+
+/**
+ * Makes a store like a WeakMapStore except that Promises and Vows can also be
+ * used as keys.
+ * NOTE: This depends on promise identity being stable!
+ *
+ * @param {string} name
+ */
+const makeVowishStore = name => {
+  // The vowMap would be needed if we supported enumeration,
+  // in order to reconstruct the original keys.
+  // const vowMap = new Map();
+  const map = new WeakMap();
+
+  return Far(name, {
+    init: (k, v) => {
+      const k2 = toPassableCap(k);
+      !map.has(k2) ||
+        // separate line so I can set a breakpoint
+        Fail`key already bound: ${k} -> ${map.get(k2)} vs ${v}`;
+      map.set(k2, v);
+    },
+    has: k => map.has(toPassableCap(k)),
+    get: k => {
+      const k2 = toPassableCap(k);
+      map.has(k2) ||
+        // separate line so I can set a breakpoint
+        Fail`key not found: ${k}`;
+      return map.get(k2);
+    },
+  });
+};
+
+/** @typedef {ReturnType<makeVowishStore>} VowishStore */
+
+/**
+ * @param {Zone} zone
+ */
+export const prepareWeakBijection = zone => {
+  /** @type {Ephemera<WeakBijection, VowishStore>} */
+  const g2h = makeEphemera(() => makeVowishStore('guestToHost'));
+  /** @type {Ephemera<WeakBijection, VowishStore>} */
+  const h2g = makeEphemera(() => makeVowishStore('hostToGuest'));
```

Comment:
> I will file an issue, this can be done as a follow-up

So resolving here.

Reply:
Filed #9365

Wondering if we should not change `WeakMap` to `Map` in the bijection for now. It would have the same memory profile, but would avoid the liveslots overhead of tracking vrefs.